compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp

```diff
@@ -631,15 +704,66 @@ StructuredOpMatcher m_StructuredOp() {
   return StructuredOpMatcher::create<OpType...>();
 }
 
+class OptionalStructuredOpMatcher;
+OptionalStructuredOpMatcher m_OptionalStructuredOp(
+    StructuredOpMatcher &&nested);
+
+/// A wrapper around a structured op matcher that injects optionality. The
+/// wrapped predicates will get applied for capturing purposes, but the `match`
+/// of this class will always succeed to allow the matching to continue. One can
+/// check if the wrapped match actually succeeded by checking if the operation
+/// was captured.
+class OptionalStructuredOpMatcher {
+  friend OptionalStructuredOpMatcher m_OptionalStructuredOp(
+      StructuredOpMatcher &&nested);
+  explicit OptionalStructuredOpMatcher(StructuredOpMatcher &&nested)
+      : nested(nested) {}
+
+ public:
+  /// Returns true if the wrapped matcher succeeded as indicated by the capture.
+  bool succeeded() const { return nested.getCaptured(); }
+
+  /// Runs the wrapped matcher and returns true regardless of its result.
+  bool match(Operation *op) {
+    (void)nested.match(op);
+    return true;
+  }
+
+ private:
+  /// Actual structured matcher.
+  StructuredOpMatcher nested;
+};
+
+/// Makes the given structured op matcher optional.
+OptionalStructuredOpMatcher m_OptionalStructuredOp(
+    StructuredOpMatcher &&nested) {
+  return OptionalStructuredOpMatcher(std::move(nested));
+}
+
+/// Creates an optional matcher for ops with the kinds provided as template
+/// arguments.
+template <typename... OpType>
+OptionalStructuredOpMatcher m_OptionalStructuredOp() {
+  return m_OptionalStructuredOp(m_StructuredOp<OpType...>());
+}
+
 }  // namespace
 
 static constexpr unsigned cudaWarpSize = 32;
 
 /// Matcher.
 static bool matchGPUReduction(linalg::LinalgOp op, TileSizesListType &tileSizes,
-                              std::array<int64_t, 3> &workgroupSize) {
+                              std::array<int64_t, 3> &workgroupSize,
+                              bool &hasTrailingEltwise) {
   // TODO: match the sequence the strategy supports.
   auto fill = m_StructuredOp<linalg::FillOp>();
+  auto trailingEltwise = m_StructuredOp<linalg::GenericOp>()
```

Comment:
nit: combine with line below into a single:

```
auto maybeTrailingEltwise = m_StructuredOp<linalg::GenericOp, Optional>()...
```

Reply:
This will require at least the code below, plus some amount of template logic to drop the last argument from the type pack. Are you sure you want this?

```c++
template <typename T>
struct MatcherConstructor {};

template <>
struct MatcherConstructor<StructuredOpMatcher> {
  template <typename... OpType>
  static StructuredOpMatcher doit() {
    return StructuredOpMatcher::create<OpType...>();
  }
};

template <>
struct MatcherConstructor<OptionalStructuredOpMatcher> {
  template <typename... OpType>
  static OptionalStructuredOpMatcher doit() {
    return m_OptionalStructuredOp(StructuredOpMatcher::create<OpType...>());
  }
};

/// Creates a matcher of a structured op with kinds provided as template
/// arguments.
template <typename... OpType>
using MaybeOptionalStructuredOpMatcher = std::conditional_t<
    std::is_same_v<std::remove_reference_t<std::tuple_element<
                       sizeof...(OpType) - 1, std::tuple<OpType...>>>,
                   OptionalMatch>,
    OptionalStructuredOpMatcher, StructuredOpMatcher>;

template <typename... OpType>
MaybeOptionalStructuredOpMatcher<OpType...>
m_StructuredOp() {
  return MatcherConstructor<MaybeOptionalStructuredOpMatcher<OpType...>>::template doit<OpType...>();
}
```

Reply:
let's not force it for now and reevaluate later, thanks for investigating!